### ISSUE TITLE
Getting corresponding location for player

### ISSUE THREAD

**Issue by kdheepak** [tfausak/rattletrap#82]
Hi @tfausak, great job with `rattletrap`. It has been able to parse every replay I've thrown at it (Windows+Mac). I'm having difficulty understanding the format though. Specifically, at this point I'm trying to get locations of the different cars at various times in the game. 

```python
frames = data["content"]["body"]["frames"]

key_frames = [f for f in frames if f["replications"] != []]

location = {}
for frame in key_frames:
    for r in frame["replications"]:
        updated = r["value"].get("updated")
        if updated is not None:
            for u in r["value"]["updated"]:
                
                if u["value"].get("rigid_body_state") is not None and u["name"] == "TAGame.RBActor_TA:ReplicatedRBState":
                    _loc = u["value"]["rigid_body_state"]["location"]
                    loc = {}
                    loc["x"] = _loc["x"]
                    loc["y"] = _loc["y"]
                    loc["z"] = _loc["z"]
                    loc["name"] = r["actor_id"]["value"]
                    location[frame["time"]] = loc
```

Firstly, I'm not sure what a `replication` is? I'm assuming it is the various events that occur in a frame, and they can be either `spawned`, `updated` or `detroyed` events. I'm also assuming that an `updated` "replication" in a frame that has the name `"TAGame.RBActor_TA:ReplicatedRBState"` contains a location. How do I find out which player (/ ball?) the different `updated` replications correspond to?

```python
array([ 31,  28,  15,   0,   9, 126, 105, 129, 127, 128,  69,  47,  68,
       131, 130, 132,  91,  84, 133,  85, 194, 171, 193, 195, 192,  64,
        46,  63])
```

If I get every `updated` replication that has a name of `"TAGame.RBActor_TA:ReplicatedRBState"`        , then the above is the unique set of `actor_id` values (`r["actor_id"]["value"]`) from each replication. This is a 2v2 game replay, so I'm assuming there's 4 players and the ball. What are the other locations in here? 

This may be a dumb question but I feel like I'm missing something (silly?) and any help will be appreciated.

**Comment by tfausak** [comment]
Hi @kdheepak! Not a dumb question at all. The replay format is difficult to understand. I'd suggest starting with a short 1v0 replay like [this one](
https://github.com/tfausak/rattletrap/raw/9342b2a0b32785dbb807ef1da6618f73e98780d7/replays/29f5.replay) because there's a lot less going on to keep track of. 

You can think of replications as new information about the game. The car a player is driving, the position of the ball, and the time left in the match are all represented as replications. The `ReplicatedRBState` replicated attribute that you're looking at is the right one. It tells you the position (and velocity) of some "rigid body" (RB) in the game. 

Notably, players are replicated separately from their cars. Every time a car is demolished, it is "destroyed" and a new one is "spawned". But the player is still the same. The game handles this by having the cars point to the players that they belong to. 

Here are some facts about the replay I linked to as an example:

- Actor ID 0 is the ball. Its class name is `TAGame.Ball_TA` and its object name is `Archetypes.Ball.Ball_Default`. It's the very first thing replicated in the replay. 
- Actor ID 1 is the only player in the replay. Its class name is `TAGame.PRI_TA` ("PRI" means "player replication info") and its object name is `TAGame.Default__PRI_TA`. It's the second thing replicated in the replay. 
- Actor ID 12 is the player's car. Its class name is `TAGame.Car_TA` and its object name is `Archetypes.Car.Car_Default`. 
- The first updated replication is for actor 0 (the ball) and it tells you its position. 
- The second updated replication is for actor 1 (the player) and it tells you a lot of information, like ping, name, team, and so on. 
- Eventually we get to an updated replication for actor 12, the player's car. That replication has a `Engine.Pawn:PlayerReplicationInfo` attribute that holds the actor ID of the associated player. In this case, that value is `1`.

Long story short, look for an updated replication like this:

``` js
      {
        "actor_id": {
          "limit": 1023,
          "value": 12 // this is the actor ID of the car
        },
        "value": {
          "updated": [
            {
              "id": {
                "limit": 68,
                "value": 30
              },
              "name": "Engine.Pawn:PlayerReplicationInfo",
              "value": {
                "flagged_int": {
                  "flag": true,
                  "int": 1 // this is the actor ID of the player
                }
              }
            },
```

I hope that helps!

**Comment by kdheepak** [comment]
I think I got it to work! Here's a heatmap I generated from X and Y location data

![heatmap](https://user-images.githubusercontent.com/1813121/43047593-c06612b8-8d96-11e8-8df2-85f5cb092937.png)

Thanks for your help!

**Comment by tfausak** [comment]
I'm glad you got it to work! 

In general you cannot count on actor IDs corresponding to particular entities. I think the ball is often actor 0, but I don't think it _always_ is. Checking the names is the only way to be sure.

**Comment by tfausak** [comment]
I don't think the map size is encoded in the replay. The header has a `MapName` property. You could in theory keep a lookup table from map names (like `Stadium_p`) to their dimensions.

**Comment by kdheepak** [comment]
I see it now, thanks!
